src/libopensc/card-iasecc.c

```diff
@@ -1245,10 +1246,17 @@ iasecc_process_fci(struct sc_card *card, struct sc_file *file,
 	for (ii = 0; ii < 7; ii++, mask /= 2)  {
 		unsigned char op = file->type == SC_FILE_TYPE_DF ? ops_DF[ii] : ops_EF[ii];
 
+		/* avoid any access to acls[offs] beyond the taglen */
+		if (offs >= taglen) {
+			if (taglen == 0)
+				sc_log(ctx, "Warning: ACL is empty, it is valid but surprising");
```

Comment:
I think this log should show up also if `taglen != 0` -- this means inconsistency in the returned data (the first byte is not matching the content bytes following if I understand what is done here). It is probably up to you whether it should be just log or error (i am fine with both).

Reply:
No, other lengths are valid. It is tight to the mask. For instance, when the length is 6, the mast is 7A, see the logs from the issue #2262

Reply:
Other lengths are valid, but only with valid mask! But here, we are catching buffer overruns, which are not valid. If we have other length, lets say 4 and mask indicating the length should be 6, we end here too, which is error and should be logged (at least).

If you have valid length and valid mask, you should not be ending here, given that you move this block after the block continue statement below, which makes sure the mask is taken into the effect.

Reply:
I got your point, see: https://github.com/vjardin/OpenSC/commit/fed485b723c436e1ec08d4410b6149feb9e61294

Reply:
I had in my mind something simpler like this:
```
		if (!(mask & acls[0])) {
			continue;
		}
		/* avoid any access to acls[offs] beyond the taglen */
		if (offs >= taglen) {
			sc_log(ctx, "Warning: Invalid offset reached during ACL parsing");
			break;
		}
```
The first part is still safe with invalid `offs` and simplifies the conditions in the second block. But your version should work the same way if I see right.

Reply:
I agree, I prefer your suggestion, it avoids too many if() else logic. See https://github.com/vjardin/OpenSC/commit/3b2581587de1c4f5007c12def8568af3c760aa4d